# Genomes :dna:

Using the most recent major releases of the human and mouse genomes  from [NCBI Datasets](https://www.ncbi.nlm.nih.gov/datasets)
|species|assembly|
|---|---|
| [human](https://www.ncbi.nlm.nih.gov/grc/human) | [GCA_000001405.15_GRCh38](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/) |
 [mouse](https://www.ncbi.nlm.nih.gov/grc/mouse) | [GCA_000001635.9_GRCm39](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/) |

## Alignment

A guide to aligning reads to a reference genome using the following tools:

- Bowtie2
- Hisat2

## Reference genomes

Using the most recent major releases of the human and mouse genomes from [NCBI Datasets](https://www.ncbi.nlm.nih.gov/datasets)

- [human](https://www.ncbi.nlm.nih.gov/grc/human)
- [mouse](https://www.ncbi.nlm.nih.gov/grc/mouse)

### Downloading the reference genomes
Index of /genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/seqs_for_alignment_pipelines.ucsc_ids

```sh
WGET_OPTS="--no-parent --no-directories --no-host-directories --no-check-certificate -e robots=off --cut-dirs=7"

wget $WGET_OPTS -r -P $HOME/genomes/mouse2/ https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/seqs_for_alignment_pipelines.ucsc_ids/

# download the checksum file at
# https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/md5checksums.txt
wget $WGET_OPTS -P $HOME/genomes/mouse2/ https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/md5checksums.txt

# Check these files
```
### Info
Check this file:
FASTA files
-----------
GCA_000001635.9_GRCm39_full_analysis_set.fna.gz

A gzipped file that contains FASTA format sequences for the following:
1. chromosomes from the GRCm39 Primary Assembly unit (C57BL/6J).
2. mitochondrial genome from the GRCm39 non-nuclear assembly unit.
3. unlocalized scaffolds from the GRCm39 Primary Assembly unit
   (C57BL/6J).
4. unplaced scaffolds from the GRCm39 Primary Assembly unit
   (C57BL/6J).

Note for Picard tool users:
Older versions of Picard tools do not support the .fna extension for
FASTA sequence files. To use the analysis_set.fna files provided in
this directory either upgrade to a newer version of Picard tools
(version 1.130 or higher) or rename the files to .fa.

> BWA index files
> ---------------
> GCA_000001635.9_GRCm39_full_analysis_set.fna.bwa_index.tar.gz
>
> tar archives containing BWA index files for the analysis sets. The BWA
> index files were generated by running bwa version 0.7.17 using the
> default parameters for the index command.
>
> Samtools index files
> --------------------
> GCA_000001635.9_GRCm39_full_analysis_set.fna.fai
>
> The Samtools index files were generated by running Samtools version
> 1.9 using the default parameters for the index command.
>
> Bowtie2 index files
> -------------------
> GCA_000001635.9_GRCm39_full_analysis_set.fna.bowtie_index.tar.gz
>
> tar archives containing Bowtie2 index files for the analysis sets. The
> Bowtie2 index files were generated by running Bowtie2 version 2.2.6
> using the default parameters for the index command.
>
> HISAT2 index files
> ------------------
> GCA_000001635.9_GRCm39_full_analysis_set.fna.hisat2_index.tar.gz
>
> tar archives containing HISAT2 index files for the analysis sets. The
> HISAT2 index files were generated by running HISAT2 version 2.0.5
> using the default parameters for the index command.
>
> Annotation files
> ----------------
> GCA_000001635.9_GRCm39_full_analysis_set.refseq_annotation.gff.gz
> GCA_000001635.9_GRCm39_full_analysis_set.refseq_annotation.gtf.gz
>
> Gzipped RefSeq annotation files in GFF3 and GTF format with column 1
> seq-ids adjusted to use the UCSC-style names to match the FASTA files.
>
> 3. Sequence names
> =================
>
> The sequence names in the analysis sets use UCSC names when these have
> already been assigned or otherwise follow the following UCSC-style
> naming patterns.

```sh

#!/bin/bash
#
# Download and verify the latest major releases of human and mouse genomes.

set -euxo pipefail

# Constants for URLs and directories
readonly NCBI_BASE_URL="https://ftp.ncbi.nlm.nih.gov/genomes/all"
readonly HUMAN_LATEST="GCA_000001405.15_GRCh38"
readonly MOUSE_LATEST="GCA_000001635.9_GRCm39"
readonly HUMAN_BASE_URL="${NCBI_BASE_URL}/GCA/000/001/405/${HUMAN_LATEST}"
readonly MOUSE_BASE_URL="${NCBI_BASE_URL}/GCA/000/001/635/${MOUSE_LATEST}"
readonly PREBUILT_INDEXES="seqs_for_alignment_pipelines.ucsc_ids"
readonly BASE_URLS=("$HUMAN_BASE_URL" "$MOUSE_BASE_URL")
readonly CHECKSUMS=(
  "${HUMAN_BASE_URL}/md5checksums.txt"
  "${MOUSE_BASE_URL}/md5checksums.txt"
)
readonly WGET_FLAGS="--quiet --show-progress --progress=bar:force:noscroll --no-parent --no-host-directories -e robots=off --cut-dirs=7"
readonly DEST_DIRS=("$HOME/genomes/human" "$HOME/genomes/mouse")

create_directories() {
  for dir in "${DEST_DIRS[@]}"; do
    mkdir -p "$dir"
  done
}

# Download prebuilt indexes recursively
download_prebuilt_indexes() {
  local prebuilt_path="$1"
  for i in "${!BASE_URLS[@]}"; do
    wget ${WGET_FLAGS} -r -P "${DEST_DIRS[$i]}" "${BASE_URLS[$i]}/$prebuilt_path" &
  done
}

# Download individual checksum files
download_checksums() {
  for i in "${!CHECKSUMS[@]}"; do
    wget ${WGET_FLAGS} -P "${DEST_DIRS[$i]}" "${CHECKSUMS[$i]}" &
 done
}

# Verify checksums
verify_checksums() {
  for dir in "${DEST_DIRS[@]}"; do
    pushd "$dir" > /dev/null

    if [ -f "md5checksums.txt" ]; then
      while read -r md5 file; do
        if ! echo "${md5} ${file}" | md5sum -c --status; then
          echo "Checksum verification failed for ${file}" >&2
          exit 1
        fi
      done < "md5checksums.txt"
    else
      echo "Checksum file not found in ${dir}" >&2
      exit 1
    fi

    popd > /dev/null
  done
}

# Main
create_directories
download_prebuilt_indexes "$PREBUILT_INDEXES"
download_checksums
#verify_checksums


```sh
# and build the index files
bowtie2-build $INDEX $PREFIX
```

```sh
bowtie2-build GCA_000001635.9_GRCm39_genomic.fna.gz GRCm39
```


curl -O https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_full_analysis_set.fna.bowtie_index.tar.gz
curl -O https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001635.9_GRCm39_full_analysis_set.fna.bowtie_index.tar.gz

## and extract the files

tar -xvf GCA_000001405.15_GRCh38_full_analysis_set.fna.bowtie_index.tar.gz
tar -xvf GCA_000001635.9_GRCm39_full_analysis_set.fna.bowtie_index.tar.gz

````
optionally, set environment variables
``` sh
export MOUSEREF=~/genomes/GCA_000001635.9_GRCm39_full_analysis_set.fna.bowtie_index
export HUMANREF=~/genomes/GCA_000001405.15_GRCh38_full_analysis_set.fna.bowtie_index
````

## Hisat2

## Bowtie2

Instead of exporting the path like `export BT2_HOME=/home/ubuntu/src/bowtie2`, install the binary by `sudo cp`ing it to `/usr/local/bin`

> **Note**: Use the `--mm` flag to use memory-mapped I/O, especially since this reduces overhead with multiple threads.
>
> see: `man mmap`

#### Building Bowtie2

Skip this section if you are not building from source.

Building from source gives the options for:

1. basic automatic dependency management and static linkage of `zstd` and `zlib`
   - `make static-libs && make STATIC_BUILD=1`
2. SRA (Sequence Read Archive) support
   - `make sra-deps && make USE_SRA=1.`
3. libsais support: a state-of-the-art suffix array construction algorithm that speeds up the building of the index
   - `[g]make libsais USE_SAIS_OPENMP=1 *`
   - this is important so it can be multithreaded

To build bowtie2-build with libsais first make sure that the libsais submodule is available. This can be done in one of the following ways:

```sh
# first time cloning
git clone --recursive https://github.com/BenLangmead/bowtie2.git

# existing checkout of bowtie2
git submodule init && git submodule update
```

#### Building with CMake

To build Bowtie2 with SRA and libsais support:

```sh
cmake . -D USE_SRA=1 -D USE_SAIS=1 && cmake --build .
```

#### Lambda phage example

from `fastq` (or the compressed `fastq.gz`) to `sam` to `bam` to `sorted.bam`

```sh
# align example paired-end reads
bowtie2 -x example/index/lambda_virus \
        -1 example/reads/reads_1.fq   \
        -2 example/reads/reads_2.fq | \
        samtools view -bS - |         \
        samtools sort > eg2.sorted.bam
```

Sorted BAM is a useful format because the alignments are

- compressed, which is convenient for long-term storage, and
- sorted, which is conveneint for variant discovery.

